osdep/macosx_touchbar.m

```diff
@@ -0,0 +1,202 @@
+    } else if ([identifier isEqualToString:cycleSubtitle]) {
+        return [self tbButtonWithId:identifier
+                               name:@"Cycle Subtitle"
+                              image:NSImageNameTouchBarComposeTemplate
+                             action:@selector(cycleSubtitle:)];
+    } else if ([identifier isEqualToString:currentPosition]) {
+        NSCustomTouchBarItem *tbItem = [self tbTextWithId:identifier name:@"Current Position"];
+        self.touchbarCurrentPosition = (NSTextField *)tbItem.view;
+        return tbItem;
+    } else if ([identifier isEqualToString:timeLeft]) {
+        NSCustomTouchBarItem *tbItem = [self tbTextWithId:identifier name:@"Time Left"];
+        self.touchbarTimeLeft = (NSTextField *)tbItem.view;
+        return tbItem;
+    }
+    return nil;
+}
+
+- (NSCustomTouchBarItem *)tbButtonWithId:(NSTouchBarItemIdentifier) identifier
+    name:(NSString *)label image:(NSString *)image action:(SEL)action
+{
+    NSCustomTouchBarItem *tbItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
+    NSImage *tbImage = [NSImage imageNamed:image];
+    tbItem.view = [NSButton buttonWithImage:tbImage target:self action:action];
+    tbItem.customizationLabel = label;
+    return tbItem;
+}
+
+- (NSCustomTouchBarItem *)tbTextWithId:(NSTouchBarItemIdentifier) identifier
+                                  name:(NSString *)label
+{
+    NSCustomTouchBarItem *tbItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
+    NSTextField *tbText = [NSTextField labelWithString:@"0:00"];
+    tbText.alignment = NSTextAlignmentCenter;
+    tbItem.view = tbText;
+    tbItem.customizationLabel = label;
+    return tbItem;
+}
+
+- (NSString *)formatTimeFrom:(int)time
+{
+    int seconds = time % 60;
+    int minutes = (time / 60) % 60;
+    int hours = time / (60 * 60);
+
+    NSString *stime = hours > 0 ? [NSString stringWithFormat:@"%d:", hours] : @"";
+    stime = (stime.length > 0 || minutes > 9) ?
+        [NSString stringWithFormat:@"%@%02d:", stime, minutes] :
+        [NSString stringWithFormat:@"%d:", minutes];
+    stime = [NSString stringWithFormat:@"%@%02d", stime, seconds];
+
+    return stime;
+}
+
+-(void)handlePropertyChange:(struct mpv_event_property *)property
+{
+    NSString *name = [NSString stringWithUTF8String:property->name];
+    mpv_format format = property->format;
+
+    if ([name isEqualToString:@"time-pos"] && format == MPV_FORMAT_DOUBLE) {
+        double position = *(double *)property->data;
+        position = position < 0 ? 0 : position;
+        if (self.duration > 0) {
+            self.touchbarSeekbar.enabled = YES;
+            if (!self.touchbarSeekbar.highlighted)
+                self.touchbarSeekbar.doubleValue = (position/self.duration)*100;
+            NSString *left = [self formatTimeFrom:(int)(floor(self.duration)-floor(position))];
+            self.touchbarTimeLeft.stringValue = [NSString stringWithFormat:@"-%@", left];
+        }
```

Comment:
duration and time-pos could be reported and enabled/disabled in random order - maybe better to move this to a function which updates them in a uniform way?